api/src/modules/scenario-interventions/scenario-interventions.service.ts

```diff
@@ -346,27 +363,58 @@ export class ScenarioInterventionsService extends AppBaseService<
 
     const newSourcingLocationData: SourcingData[] = [];
     for (const location of sourcingData) {
-      const newInterventionLocation: SourcingData = {
-        materialId: location.materialId,
-        locationType: dto.newLocationType,
-        locationAddressInput: dto.newLocationAddressInput,
-        locationCountryInput: dto.newLocationCountryInput,
-        locationLatitude: dto.newLocationLatitude,
-        locationLongitude: dto.newLocationLongitude,
-        t1SupplierId: dto.newT1SupplierId,
-        producerId: dto.newProducerId,
-        businessUnitId: location.businessUnitId,
-        geoRegionId: geoCodedLocationSample[0].geoRegionId,
-        adminRegionId: geoCodedLocationSample[0].adminRegionId,
-        locationWarning: geoCodedLocationSample[0].locationWarning,
-        sourcingRecords: location.sourcingRecords.map((elem: any) => {
-          return { year: elem.year, tonnage: elem.tonnage };
-        }),
-        interventionType: SOURCING_LOCATION_TYPE_BY_INTERVENTION.REPLACING,
-      };
+      const identicalSourcingLocationDataIndex: number | undefined =
+        newSourcingLocationData.findIndex(
+          (el: SourcingData) =>
+            location.materialId === el.materialId &&
+            location.businessUnitId === el.businessUnitId,
+        );
+
+      if (identicalSourcingLocationDataIndex >= 0) {
+        const joinedRecords: any = newSourcingLocationData[
+          identicalSourcingLocationDataIndex
+        ].sourcingRecords.concat(location.sourcingRecords);
+
+        const mergedRecords = joinedRecords.reduce(
+          (acc: any, sourcingRecords: { year: number; tonnage: number }) => {
+            acc[sourcingRecords.year] = {
+              year: sourcingRecords.year,
+              tonnage:
+                (acc[sourcingRecords.year]
+                  ? Number(acc[sourcingRecords.year].tonnage)
+                  : 0) + Number(sourcingRecords.tonnage),
+            };
+            return acc;
+          },
+          {},
+        );
+        newSourcingLocationData[
+          identicalSourcingLocationDataIndex
+        ].sourcingRecords = Object.values(mergedRecords);
```

Comment:
It would be nice to extract this to a helper function, so that the for block is less cluttered, and the intention more clear